jni/src/main/java/com/horizen/cswnative/CswProof.java

```diff
@@ -0,0 +1,342 @@
+     * @param scTxCommTreeHeight - Height of the subtree of the SCsCommitmentTree having as leaves the single ScTxsCommitments  
+     * @param provingKeyPath - file path to which saving the proving key. Proving key will be saved in compressed form.
+     * @param verificationKeyPath - file path to which saving the verification key. Verification key will be saved in compressed form.
+     * @param maxProofPlusVkSize - maximum allowed size for proof + vk, estimated assuming not to use zk property
+     * @return true if (pk, vk) generation and saving to file was successfull, false otherwise.
+     */
+    public static boolean setup(
+        ProvingSystemType psType,
+        int rangeSize,
+        int mstTreeHeight,
+        int ftTreeHeight,
+        int scTxCommTreeHeight,
+        String provingKeyPath,
+        String verificationKeyPath,
+        int maxProofPlusVkSize
+    )
+    {
+        return nativeSetup(
+            psType, rangeSize, mstTreeHeight, ftTreeHeight, scTxCommTreeHeight,
+            provingKeyPath, verificationKeyPath, false, maxProofPlusVkSize, true, true
+        );
+    }
+
+    private static native CreateProofResult nativeCreateProof(
+        CswSysData sysData,
+        FieldElement scId,
+        WithdrawalCertificate lastWcert,
+        CswUtxoProverData utxoData,
+        CswFtProverData ftData,
+        String provingKeyPath,
+        boolean checkProvingKey,
+        boolean zk,
+        boolean compressed_pk,
+        boolean compress_proof
+    );
+
+    /**
+     * Compute proof for given parameters
+     * @param sysData - certificate sys data.
+     * @param scId - the id of the corresponding sidechain
+     * @param lastWcert - the last confirmed wcert in the MC. Can be empty if SC has ceased before we have at least
+     *                    certs for 2 epochs (in this case we can only withdraw FT)
+     * @param utxoData - data required to prove withdraw of a SC utxo. It's empty if the prover wants to prove
+     *                   withdraw of a FT instead. If this field is present, then lastWCert must be present too.
+     * @param ftData - data required to prove withdraw of a FT. It's empty if the prover wants to prove
+     *                 withdraw of a SC utxo instead.
+     * @param provingKeyPath - file path from which reading the proving key
+     * @param checkProvingKey - enable semantic checks on the proving key (WARNING: very expensive)
+     * @param zk - if proof must be created using zk property or not
+     * @param compressed_pk - if the pk read from provingKeyPath is in compressed form or not
+     * @param compress_proof - whether to return the proof bytes in compressed form or not
+     * @return a CreateProofResult instance, i.e. the computed proof bytes and the quality of the certificate (i.e. in this case, number of valid signatures),
+     *         OR null pointer if some errors occured during proof creation.
+     */
+    public static CreateProofResult createProof(
```

Comment:
add `throws IllegalArgumentException`. Same in other places.

Reply:
Done